### PR TITLE
Desi fixes to delta extraction

### PR DESCRIPTION
**Pull request by Waelthus**
This is a first step to getting the desi mode working for `delta_extraction`. Still needs some testing and cleanup (e.g. for writing output names as lambda instead of loglam), but at least will run through until the end.

### CODE REVIEW

File: py/picca/delta_extraction/data_catalogues/desi_data.py [hunk 1]
```diff
@@ -307,35 +310,33 @@ def read_from_minisv_desi(self, catalogue):
...
-                night_spec = int(filename.split('-')[-1].split('.')[0])
-                colors = ['B', 'R', 'Z']
-                if index == 0:
-                    logging.warning(
-                        "Couldn't read the all band-coadd, trying "
-                        "single band as introduced in Andes reduction")
-            ra = np.radians(ra)
-            dec = np.radians(dec)
+            ra_spec = np.radians(ra_spec)
+            dec_spec = np.radians(dec_spec)
```
Comment (iprafols) [inline py/picca/delta_extraction/data_catalogues/desi_data.py:339]: We are not using ra_spec and dec_spec for anything, aren't we? Or did I miss something? Maybe we can just delete lines 319, 320, 321, 338, and 339

File: py/picca/delta_extraction/data_catalogues/desi_data.py [hunk 2]
```diff
@@ -444,5 +445,5 @@ def read_from_minisv_desi(self, catalogue):
 
         if num_data == 0:
             raise DataError("No Quasars found, stopping here")
-
+        #breakpoint()
```
Comment (iprafols) [inline py/picca/delta_extraction/data_catalogues/desi_data.py:448]: we should remove this

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 3]
```diff
@@ -197,18 +197,24 @@ def _initialize_variables_lin(self):
         # initialize the mean quasar continuum
         # TODO: maybe we can drop this and compute first the mean quasar
         # continuum on compute_mean_expected_flux
+
+        #TODO: following line is non-ideal and will lead to mean continuum bins that are wider than neccessary
+        #      as rest_frame pixels are only delta_lambda/(1+z) wide (different for each spectrum), i.e. can do
+        #      better by a factor 3 at least
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:203]: How would you change this?

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 4]
```diff
@@ -634,14 +640,17 @@ def compute_delta_stack(self, forests, stack_from_deltas=False):
                 bins = ((forest.log_lambda - Forest.log_lambda_min) /
                         Forest.delta_log_lambda + 0.5).astype(int)
             elif Forest.wave_solution == "lin":
-                bins = ((forest.lambda_ - Forest.log_lambda_min) /
-                        Forest.delta_log_lambda + 0.5).astype(int)
+                #TODO: this is definitely wrong, we do not have log-lambda quantities at this point, there might be subtleties as well
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:643]: Remove the TODO since it is now fixed

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 5]
```diff
@@ -634,14 +640,17 @@ def compute_delta_stack(self, forests, stack_from_deltas=False):
...
+
+                bins = ((forest.lambda_ - Forest.lambda_min) /
+                        Forest.delta_lambda + 0.5).astype(int)
             else:
                 raise ExpectedFluxError("Forest.wave_solution must be either "
                                         "'log' or 'linear'")
-            rebin = np.bincount(bins, weights=delta * weights)
+            #the indices are to exclude lambda>lambda_max from the stack
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:650]: Why is this necessary? Forest only contains lambda<lambda_max, doen't it?

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 6]
```diff
@@ -634,14 +640,17 @@ def compute_delta_stack(self, forests, stack_from_deltas=False):
...
+                        Forest.delta_lambda + 0.5).astype(int)
             else:
                 raise ExpectedFluxError("Forest.wave_solution must be either "
                                         "'log' or 'linear'")
-            rebin = np.bincount(bins, weights=delta * weights)
+            #the indices are to exclude lambda>lambda_max from the stack                                        
+            rebin = np.bincount(bins[bins<num_bins], weights=delta[bins<num_bins] * weights[bins<num_bins])
             stack_delta[:len(rebin)] += rebin
-            rebin = np.bincount(bins, weights=weights)
+            rebin = np.bincount(bins[bins<num_bins], weights=weights[bins<num_bins])
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:653]: same here

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 7]
```diff
@@ -694,27 +703,37 @@ def compute_mean_cont_lin(self, forests):
...
         #    1/Cont_old * <F/spectrum_dependent_fitting_function>
         for forest in forests:
+
             bins = (
                 (forest.lambda_ /
                  (1 + forest.z) - Forest.lambda_min_rest_frame) /
                 (Forest.lambda_max_rest_frame - Forest.lambda_min_rest_frame) *
                 num_bins).astype(int)
 
+            #this catches issues where the spectrum is longer than the range wanted for continuum fitting
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:713]: this is related with the two changes immediately above

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py [hunk 8]
```diff
@@ -694,27 +703,37 @@ def compute_mean_cont_lin(self, forests):
...
+            bins=bins[select_bins]
+
             var_lss = self.get_var_lss(forest.lambda_)
             eta = self.get_eta(forest.lambda_)
             fudge = self.get_fudge(forest.lambda_)
             var_pipe = 1. / forest.ivar / forest.continuum**2
             variance = eta * var_pipe + var_lss + fudge / var_pipe
             weights = 1 / variance
+
+            if np.asarray(forest.continuum).size==1:
```
Comment (iprafols) [inline py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:724]: I'm not sure I understand what this is for

File: py/picca/delta_extraction/data_catalogues/desi_data.py [hunk 9]
```diff
@@ -429,9 +428,11 @@ def read_from_minisv_desi(self, catalogue):
                                 "z": entry['Z'],
                                 "petal": entry["PETAL_LOC"],
                                 "tile": entry["TILEID"],
-                                "night": entry["NIGHT"],
+                                "night": entry[nightcol],
                                 "exposures_diff": exposures_diff,
-                                "reso": reso_in_km_per_s
+                                "reso": np.real(reso_in_km_per_s),
+                                "reso_AA": np.real(reso_in_AA),
```
Comment (iprafols) [inline py/picca/delta_extraction/data_catalogues/desi_data.py:434]: You have modified the function to return the resolution also in AA. Is there a reason for this? Do you plan to use this instead of using the resolution in km/s? If so, we should replace it accordingly, not just simply add it. Note that currently this is being ignored as it is not used by the constructor of Pk1dForest

File: py/picca/delta_extraction/utils_pk1d.py [hunk 10]
```diff
@@ -162,6 +164,7 @@ def spectral_resolution_desi(reso_matrix, lambda_):
             4.0 / 2.0 / np.log(
                 reso[len(reso) // 2][:] / reso[len(reso) // 2 - 2][:]))) / 2.0
 
-    reso_in_km_per_s = (rms_in_pixel * SPEED_LIGHT * delta_lambda)
+    reso_in_AA = (rms_in_pixel * delta_lambda)
+    reso_in_km_per_s = reso_in_AA * SPEED_LIGHT / lambda_
 
-    return reso_in_km_per_s
+    return reso_in_km_per_s, reso_in_AA
```
Comment (iprafols) [inline py/picca/delta_extraction/utils_pk1d.py:170]: See the related point in DesiData

**Review by iprafols — CHANGES_REQUESTED** [review]

File: py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py (hunk 5 above)
Comment (Waelthus) [reply py/picca/delta_extraction/expected_fluxes/dr16_expected_flux.py:650]: It might actually be that one of the earlier steps led to having us pixels with higher lambda. I'll need to double check this again.

File: py/picca/delta_extraction/data_catalogues/desi_data.py (hunk 9 above)
Comment (Waelthus) [reply py/picca/delta_extraction/data_catalogues/desi_data.py:434]: In the end there are at least 3 measures of resolution that might be informative here, the reso in km/s (varying with lambda) in AA (roughly constant with lambda) and the actual resolution matrix of the pixels (with masking applied). Depending on how we do P1d each of them could be used in principle. I guess it makes sense to have both available for summaries (e.g. mean reso), but the native one for DESI would be reso_AA.

File: py/picca/delta_extraction/utils_pk1d.py (hunk 10 above)
Comment (Waelthus) [reply py/picca/delta_extraction/utils_pk1d.py:170]: Here I would definitely recommend to just print out both for flexibility. It's a relatively crude estimate anyway.

File: py/picca/delta_extraction/data_catalogues/desi_data.py (hunk 1 above)
Comment (iprafols) [reply py/picca/delta_extraction/data_catalogues/desi_data.py]: These lines are removed in commit #cc64256

File: py/picca/delta_extraction/data_catalogues/desi_data.py (hunk 2 above)
Comment (iprafols) [reply py/picca/delta_extraction/data_catalogues/desi_data.py]: This is removed in commit #cc64256

**Comment by Waelthus** [comment]
this is now stale enough to never be merged, I'll leave it open for while we're debugging the delta_extraction branch and then will close without merging.

**Comment by iprafols** [comment]
I agree, I added the wontfix flag so that we remember

**Comment by Waelthus** [comment]
closing this as current version of the branch does run with linear bininng now. So this branch won't help anymore...
